Recovery optimization

mengranwo opened the pull request:
**Issue:**
https://github.com/apple/foundationdb/issues/53

**Current solution:**
Based on branch [release-6.0](https://github.com/apple/foundationdb/tree/release-6.0)
During recovery process for memory storage engine, we’ll check periodically whether cluster is healthy or not (by calling StatusClient::statusFetcher(<file>), I believe this is the only way to check status). If it’s healthy then abort recovery and join in cluster as an empty node. Otherwise, keeping the recovery process going until it’s done. 

**Testing:**
We're currently setting up testing env using onecloud(VMware compute resources)
See jira [ticket](https://wavefront.atlassian.net/browse/ESO-940)

mengranwo committed:
checking cluster is healthy or not during recovery process(for storage engine), if healthy, delete data files and join as new

etschannen committed:
update versions target to 6.0.18

etschannen committed:
update installer WIX GUID following release

etschannen committed:
Merge pull request #968 from etschannen/post-release-cleanup-6.0.17

Post release cleanup 6.0.17

mengranwo commented:
@panghy 
@jbau 
@warshawd

panghy commented:
Awesome, wtal

panghy commented on fdbserver/KeyValueStoreMemory.actor.cpp:
```diff
@@ -427,7 +429,34 @@ class KeyValueStoreMemory : public IKeyValueStore, NonCopyable {
 
-	ACTOR static Future<Void> recover( KeyValueStoreMemory* self, bool exactRecovery ) {
+	// reset everything and join as an empty node
+	static void join_as_empty(KeyValueStoreMemory* self){
```
this should not be just for memory storage, there's a particular `throw please_reboot_delete();` that's designed to recycle the SS and purge all data.

panghy commented on fdbserver/KeyValueStoreMemory.actor.cpp:
```diff
@@ -427,7 +429,34 @@ class KeyValueStoreMemory : public IKeyValueStore, NonCopyable {
+	// check if cluster is healthy, private method used by recovery process
+	ACTOR static Future<bool> is_cluster_healthy(){
+		// TODO: user can specify cluster file name
+		state std::pair<std::string, bool> resolvedClusterFile = ClusterConnectionFile::lookupClusterFileName("");
```
hm, this is not really correct since the cluster file can be different for the process (and not "", which resolves to /etc/foundationdb/fdb.cluster by default on linux). There should be a way to do this over the network (to fetch the recovery status of the cluster. Please dig and ping me if you can't find it.

panghy commented on fdbserver/KeyValueStoreMemory.actor.cpp:
```diff
@@ -546,13 +575,21 @@ class KeyValueStoreMemory : public IKeyValueStore, NonCopyable {
 				if (loggingDelay.isReady()) {
+					state bool isHealthy = wait(is_cluster_healthy());
 					TraceEvent("KVSMemRecoveryLogSnap", self->id)
+					    .detail("IsHealthy", isHealthy)
```
the code base has odd tabs and spacing, you need to basically follow what's already there

panghy commented on fdbserver/KeyValueStoreMemory.actor.cpp:
```diff
@@ -546,13 +575,21 @@ class KeyValueStoreMemory : public IKeyValueStore, NonCopyable {
 						.detail("Commits", dbgCommitCount)
 						.detail("EndsAt", self->log->getNextReadLocation());
-					loggingDelay = delay(1.0);
+					if(isHealthy){
```
formatting (space before "(")

mengranwo commented:
I've made changes based on @panghy's comments. 
Two major things in total: 
-  add loop inside **storage server** reboot constructor, checking every 2 secs whether (1) KVStore is completed with recovery process  (2) cluster is healthy 
- If cluster is healthy, then throw please_reboot() request, **worker** layer will take care of it automatically. 

@warshawd

Commit:
Rewrote backup metadata scheme to fix several design flaws involving cancelled or concurrent operations.  Most importantly, before an expire deletes any data it marks the end of the range being deleted as 'unreliable' so further reads of the backup will treat versions before that point as having data holes.

Commit:
Added backup URL and exceptions to trace events where applicable.

Commit:
Backup describe's "deep scan" mode should only ignore log begin/end versions, not expire and unreliable end versions.

Commit:
Bug fix, metadata read futures were moved to a vector but then not waited on.

Commit:
Static de-initialization order isn't well thought out so use flushAndExit() instead of returning from main after runNetwork() has been called.

Commit:
Backup describe can now analyze a backup assuming log data starts at a given version.  This is used by expire both as an optimization and because doing so enables expire to repair a bad metadata state that can result from a cancelled or failed expire operation from fdbbackup <= 6.0.17.  ListLogFiles() and ListRangeFiles() no longer sort results as in most cases the sort is not required and the result set can be very large.  Describe will now update minLogBegin metadata to the beginning of the log range known to be present when possible.   Several serial log and range list pairings are now done in parallel.

Commit:
Added fdbbackup expire options to calculate approximate version boundaries based on a number of days prior to the latest log file found in the backup container.  This enables expiration operations based on time (with reasonable precision) without accessing the source cluster.

Commit:
Bug fix in backup describe, do not update log begin/end metadata in backup if describe was given a start version override as it can result in incorrect metadata.

Commit:
Backup expire will fail earlier if force option is needed but not specified.

Commit:
Error code was misleading, added comment.

Commit:
In backup describe, when not using the original cluster to resolve versions to date/time strings the versions will be converted to approximate day deltas from maxLogEndVersion (if available) using core_versionspersecond.

mengranwo commented:
Push first working version. Test on FDB distributed performance workload.(70 % write and 30 % read, lasts for 20 mins)

**TODO:** 

- Currently, if the returned cluster status is healthy, I only log but never join cluster as an empty node. 

- If ss hang there for 2 mins and repeatedly query cluster status, the return value could be unstable sometimes. Don't know for sure the reasons behind.

Commit:
Added progress reporting to backup expiration.  Simplified backup delete progress.

Commit:
Print a final deleted object count at the end of backup deletion.

Commit:
Backup describe, expire, and delete will now clearly indicate when there is no backup at the given URL.

Commit:
Backup dumpFileList() can now take a version range.

Commit:
Added fdbbackup dump option to list files and sizes in a backup for a given version range.

Commit:
Added --loggroup option to all fdbbackup modes.

Commit:
Stopped escaping some printable characters in urlEscape() method for normalizing HTTP resource strings.

Commit:
Stop escaping dot in HTTP resource string normalization.

etschannen committed:
byte sample recovery did not use limits for its range reads, leading to slow tasks

Commit:
Bug fix in blobstore request loop where zero-content requests (i.e. not uploads) could cause mismatches between requests and responses.  Added HTTPRequest debug event.

etschannen committed:
Merge pull request #1007 from satherton/fix-backup-expire-bugs

Backup fixes and improvements

alexmiller-apple committed:
Merge pull request #1001 from etschannen/release-6.0

byte sample recovery did not use limits for its range reads, leading to slow tasks

Commit:
Added X-Request-ID header to HTTP requests and verification of matching ID in response, if present.

etschannen committed:
updated documentation for 6.0.18

etschannen committed:
updated release notes

etschannen committed:
fixed a typo

etschannen committed:
Merge pull request #1009 from etschannen/release-6.0

Updated documentation for 6.0.18

ajbeamon committed:
Fix: fast allocator would not cleanup memory for a thread if that thread never called getMagazine. This could happen if the first thing the thread did was to release memory.
Added a new metric for the number of threads that hold memory for each size and improve some existing metrics.
Fix: a failed ASSERT would crash if done early in the program lifetime.

ajbeamon committed:
Add PR to release notes.

Commit:
Added optional support for sending a unique id per request in a header for logging/tracking purposes.

etschannen committed:
Merge pull request #1011 from ajbeamon/fix-client-mem-leak

Fix: fast allocator would not cleanup memory for a thread if...

Commit:
When starting a restore, if range files are missing their names will be logged before an error is thrown.  The error thrown is now restore_missing_data instead of restore_corrupted_data.

Commit:
Updated unit test to new backup container behavior.

Commit:
Updated backup correctness test to new behavior.  WaitBackup() can now return the UID and BackupContainer atomically with the status code for a backup tag.

Commit:
Restore start with dry run option no longer requires or uses a database.  Removed duplicate and unnecessary code.  Dry run execution is more clear, it does not pretend to have actually restored data.  Fixed typo in help text.

Commit:
Bug fix, restore start dry run would not open a trace file.

etschannen committed:
Merge pull request #1019 from satherton/http-request-id

Backup usability and sanity check improvements

etschannen committed:
update installer WIX GUID following release

etschannen committed:
Merge pull request #1022 from etschannen/prepare-release-6.0.18

update installer WIX GUID following release

etschannen committed:
update versions target to 6.0.19

etschannen committed:
update installer WIX GUID following release

etschannen committed:
Merge pull request #1023 from etschannen/post-release-cleanup-6.0.18

Post release cleanup 6.0.18

ajbeamon committed:
Go bindings reported the wrong required version when loading an incompatible version of fdb_c.

alecgrieser committed:
Merge pull request #1053 from ajbeamon/fix-go-unsupported-version-error

Go bindings reported the wrong required version when loading an incompatible

sfc-gh-mpilman committed:
Fix bug that in `include` command

This bug results in `include` to include all
machines that have a prefix of a given machine.

This is a weird bug that has to do with ASCII
ordering. Instead this command executes two
clears: one for the IP address and one for
all IP:PORT pairs

AlvinMooreSr committed:
Added build services within docker compose

mpilman committed:
Remove duplicate comment

sfc-gh-mpilman committed:
Added release notes for `include` bug

ajbeamon committed:
Wording

Co-Authored-By: mpilman <markus@pilman.ch>

ajbeamon committed:
Fix embarrassing typo

Co-Authored-By: mpilman <markus@pilman.ch>

ajbeamon committed:
Merge pull request #1121 from mpilman/features/fdbcli-include-r6

Fix bug in `include` command

AlvinMooreSr committed:
Added support for python3 and cmake

AlvinMooreSr committed:
Added support for building binaries package

AlvinMooreSr committed:
Added support for building via cmake when present via docker-compose
Added support for running ctests via docker-compose

brownleej committed:
Merge pull request #1134 from AlvinMooreSr/make-bins

Added Binaries Package Target

AlvinMooreSr committed:
Added support for a type of correctness service run against the fdbserver binary via docker-compose

AlvinMooreSr committed:
Added support for variable build directory

ajbeamon committed:
Add release note for previous Go bindings fix.

AlvinMooreSr committed:
Merge pull request #1152 from AlvinMooreSr/docker-correctness

Server Correctness Via Docker-compose

etschannen committed:
Merge pull request #1154 from ajbeamon/release-6.0

Add release note for previous Go bindings fix.

Commit:
add go.mod

alecgrieser committed:
Merge pull request #1179 from dangerousvasil/release-6.0

add go.mod

Commit:
Bug fix in AsyncFileReadAhead where a file size that is an integer multiple of the read chunk size will cause a crash when reading the file's final block.  BackupContainerLocalDirectory now uses AsyncFileReadAhead in simulation to get simulation coverage of that class, and FileBackup will generate file sizes which expose the bug.

xumengpanda committed:
Merge pull request #1205 from satherton/fix-read-ahead-bug

Bug fix in AsyncFileReadAhead affecting FileRestore

Commit:
Added blob store Backup URL parameter 'header' which enables addition of custom HTTP header fields to blob store HTTP requests.  Added 'fdbbackup modify' command line tool for changing the backup URL and parameters, default snapshot interval, and/or current snapshot interval of a running backup.

Commit:
Added help and documentation for fdbbackup modify.  Add BackupUID and more parseable BackupURL to fdbbackup status output.

Commit:
More documentation and release note changes for fdbbackup modify.  Added release note for #1205 bug fix which was missed.

Commit:
An HTTP RequestID/ResponseID mismatch is logged as only a warning if the response code is 5xx and the responseID is empty.

Commit:
Bug fixes in fdbbackup modify command.  Active snapshot interval was wrong.  New backup destination containers were not being created.  Renamed some variables for clarity.

Commit:
Bug/usability fix:  The output URLs of fdbbackup list were meant to be directly usable for backup management operations but they were missing the bucket URL parameter.

Commit:
Fdbbackup binary will ignore .debug suffix when determining how it was invoked.

Commit:
Bug fix: fdbbackup modify on a tag that has never been used before would result in an internal error.

alecgrieser committed:
Resolves #1235: Java: FDBExceptions are created successful operation completion (#1236)

The native function `NativeFuture::Future_getError` now returns `null` when the error code is 0 instead of an `FDBException` with a "Success" message and an error code of 0. This was only used in two places within the codebase; those two places now check for `null` errors and treats them like successes.

alecgrieser committed:
Merge pull request #1256 from alecgrieser/01235-release-6.0-cherry-pick

Cherry pick #1236 onto release-6.0

Commit:
Merge branch 'release-6.0' into fdbbackup-modify

Commit:
Remove short option name for active snapshot interval

Commit:
Update documentation for modify active snapshot interval option

apkar committed:
Merge pull request #1237 from satherton/fdbbackup-modify

Fdbbackup modify command and support for custom HTTP header fields
